pre_commit/languages/coursier.py

```diff
@@ -0,0 +1,65 @@
+import contextlib
+import os
+from typing import Sequence, Generator
+from typing import Tuple
+
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import Var
+from pre_commit.envcontext import PatchesT
+from pre_commit.hook import Hook
+from pre_commit.languages import helpers
+from pre_commit.prefix import Prefix
+from pre_commit.util import clean_path_on_failure
+
+ENVIRONMENT_DIR = 'coursier'
+
+get_default_version = helpers.basic_get_default_version
+healthy = helpers.basic_healthy
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+    helpers.assert_version_default('coursier', version)
+    helpers.assert_no_additional_deps('coursier', additional_dependencies)
+
+    envdir = prefix.path(helpers.environment_dir(ENVIRONMENT_DIR, version))
+    with clean_path_on_failure(envdir):
+        os.mkdir(envdir)
+        helpers.run_setup_cmd(
+            prefix,
+            (
+                'coursier',
+                'install',
+                '--default-channels=false',
+                '--channel=.pre-commit-channel',
+                'pre-commit-hook',
+                f'--dir={envdir}'
+            )
+        )
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix) -> Generator[None, None, None]:
+    target_dir = prefix.path(
+        helpers.environment_dir(ENVIRONMENT_DIR, get_default_version()),
+    )
+    with envcontext(get_env_patch(target_dir)):
+        yield
+
+
+def run_hook(
+        hook: Hook,
+        file_args: Sequence[str],
+        color: bool,
+) -> Tuple[int, bytes]:
+    with in_env(hook.prefix):
+        return helpers.run_xargs(hook, ('pre-commit-hook',), file_args, color=color)
```

Comment:
I think the hardcoded string `'pre-commit-hook'` needs to be used with the normal `hook.entry` / `hook.cmd` stuff? not sure

Reply:
This is very similar to the docker integration and I don't think `hook.entry` or `hook.cmd` have meaning there.

I'll add an example repo under `testing/` to clarify and iron this out.

Reply:
ah yeah, the docker integration still utilizes `entry` in case users wanted to set arguments there: https://github.com/pre-commit/pre-commit/blob/202f0bbbc9c921790efdb8709e2bb961cdf7b0c3/pre_commit/languages/docker.py#L90-L93

can probably do the same as that but replace the first segment with the hardcoded "container" name

Reply:
Hmmm, looks like entry does have meaning for the docker integration. Let me think this through a bit :)